### PR TITLE
Fix race condition around stale parameters

### PR DESCRIPTION
**Pull request by GDYendell**
- Append new parameters to existing instead of overwriting them
- Add locking around accessing _stale_parameters
- Add blocking option to fetch_stale_parameters to ensure everything updated
  before starting an acquisition

### CODE REVIEW

File: control/eiger/eiger_detector.py [hunk 1]
```diff
@@ -538,7 +540,13 @@ def write_detector_config(self, item, value):
...
         if item == 'stale_parameters':
             # Read stale parameters flag
-            response = {u'value': len(self._stale_parameters) != 0, u'access_mode': u'r', u'value_type': u'bool'}
+            with self._lock:
+                response = {
+                    u'value': len(self._stale_parameters) != 0,
+                    u'access_mode': u'r',
+                    u'value_type':
+                    u'bool'
+                }
```
Comment (GDYendell) [inline control/eiger/eiger_detector.py:549]: I don't think we need this bit when we have the the set at line 717 as well. Am I missing something @ajgdls ?

File: control/eiger/eiger_detector.py [hunk 2]
```diff
@@ -682,15 +690,25 @@ def initialize_detector(self):
...
             logging.info("Read from detector [{}]: {}".format(cfg, param))
             setattr(self, cfg, param)
-        self._stale_parameters = []
+
+        # _stale_parameters may have been repopulated again by the time we get here and
+        # the flag will stay set - if so we fetch again here or in the next status poll
+        if blocking:
+            logging.info("Stale parameters repopulated - fetching again")
+            self.fetch_stale_parameters(blocking)
+
```
Comment (ajgdls) [inline control/eiger/eiger_detector.py:711]: Have I misunderstood this piece of code.  If blocking=True wouldn't you end up in an endless recursive loop?

Comment (GDYendell) [reply control/eiger/eiger_detector.py:711]: Only if puts are continuously sent that mean things need to be fetched.

Comment (GDYendell) [reply control/eiger/eiger_detector.py:711]: Sorry you are right. How about now?

Comment (ajgdls) [reply control/eiger/eiger_detector.py:711]: OK then.

**Review by ajgdls — APPROVED** [review]

File: control/eiger/eiger_detector.py [hunk 3]
```diff
@@ -682,15 +690,25 @@ def initialize_detector(self):
...
+        # _stale_parameters may have been repopulated again by the time we get here and
+        # the flag will stay set - if so we fetch again here or in the next status poll
+        if blocking and self._stale_parameters:
+            logging.info("Stale parameters repopulated - fetching again")
+            self.fetch_stale_parameters(blocking)
+
         self.update_stale_parameters()
 
     def update_stale_parameters(self):
```
Comment (GDYendell) [inline control/eiger/eiger_detector.py:714]: Here it sets the value in the parameter tree, which presumably mean GETs would receive the correct value. But, at 541 it seems to intercepts the requests for it and manual constructs the response. Maybe they don't do the same thing.

File: control/eiger/eiger_detector.py (hunk 1 above)
Comment (GDYendell) [reply control/eiger/eiger_detector.py:549]: I added another comment near the line. It is 699 pre diff and 717 including the diff

**Pull request closed without merging**